rt/x86-64-linux/System.v3

```diff
@@ -8,7 +8,120 @@
 // XXX: use on-stack storage for small IO buffers
 // XXX: what about GC moving arrays during IO operations?
 // XXX: merge common code with darwin/System.v3 into a common Posix implementation
+
+// XXX: Unsound as we need to enforce that ipv4 and ipv6 are exactly 4 and 16
+// u8s long respectively. One way to achieve this could be to create two more
+// types, Inet4Array and Inet6Array, like so:
+// class Inet4Array {
+// 	def addr: Array<u8> = Array.new(4);
+// }
+// class Inet6Array {
+// 	def addr: Array<u8> = Array.new(16);
+// }
+// type InetAddress {
+// 	case Inet4(addr: Inet4Array);
+// 	case Inet6(addr: Inet6Array);
+// }
+/**
+ * Represents an internet address. It can be one of Inet4 for IPv4 addresses or
+ * Inet6 for IPv6 addresses.
+ */
+type InetAddress {
+	case Inet4(addr: Array<u8>);
+	case Inet6(addr: Array<u8>);
+}
+
+class SocketBase {
+	// IP address of socket
+	def addr: InetAddress;
+	// Associated file descriptor of socket
+	var fd: int;
+	// Local port the socket is connected to
+	def localPort: int;
+	// Server port the socket is connected to
+	def port: int;
+
+	new(addr, localPort, port) {
+		var family: int;
+		match (addr) {
+			Inet4(addrV4) => family = LinuxConst.AF_INET;
+			Inet6(addrV6) => family = LinuxConst.AF_INET6;
+		}
+		fd = System.socket(family, LinuxConst.SOCK_STREAM /* type */, 0 /* protocol */);
+	}
+}
+
+class ClientSocket extends SocketBase {
+	new(addr: InetAddress, localPort: int, port: int) super(addr, localPort, port) { }
+
+	def connect() -> int {
+		match (addr) {
+			Inet4(addrV4) => {
+				// Definition of sockaddr for IPv4:
+				// struct sockaddr_in {
+				//   short int          sin_family;  // Address family, AF_INET
+				//   unsigned short int sin_port;    // Port number
+				//   struct in_addr     sin_addr;    // Internet address
+				//   unsigned char      sin_zero[8]; // Same size as struct sockaddr
+				// };
+				def sockAddr = Array<byte>.new(LinuxConst.SOCKADDR_V4_SIZE);
+				def nboPort = u16ToNetworkByteOrder(u16.view(port));
+				def nboAddr = u32ToNetworkByteOrder(
+						u32.view((int.view(addrV4[0]) << 24) | (int.view(addrV4[1]) << 16)
+							| (int.view(addrV4[2]) << 8) | addrV4[3])
+				);
+				// XXX: Disgusting hack alert! Hopefully this can be cleaned up with structs
```

Comment:
Another way to do this is with a `DataWriter`.

Reply:
Done. Thanks!